### PR TITLE
fallback to cvmfs if xrdcp/xrdfs not in path

### PR DESCRIPTION
**Pull request by hufnagel**
Some sites don't seem to have the xrootd client in their environment. Add a fallback to use the xrootd client from the CMSSW cvmfs repo. I am using that repo since it's the only one that is universally available.

Unfortunately we can only use an outdated xrootd version this way, since newer versions also require updated libraries that are only available in newer CMSSW builds and not in a native SL6 environment.

@bbockelm , you see any problems with using xrootd 3.3.3 as fallback if a site doesn't have xrdcp/xrdfs in the path ? This should only effect LogCollect stageout to CERN EOS and Castor since if a site chooses xrdcp for it's local stageout method, they should have a newer xrootd client deployed locally.

### CODE REVIEW

**Comment by hufnagel** [comment]
Tested the code path that finds the local xrdcp/xrdfs. Still testing using the xrootd client from cvmfs.

**Comment by hufnagel** [comment]
Ok, this doesn't work.

Plugin No such file or directory loading sec.protocol libXrdSecgsi.so
xrdcp: Copy to eoscms-srv-m2.cern.ch failed on open!
xrdcp: Unable to give access - user access restricted - not authorized identity used ; Permission denied
Plugin No such file or directory loading sec.protocol libXrdSecgsi.so
[ERROR] Server responded with an error: [3010] Unable to give access - user access restricted - not authorized identity used ; Permission denied

This 3.3.3 version can't authenticate to EOS. Which means this approach won't work unless the even older 3.2.4 version for some reason works. But I don't even want to try that.

@amaltaro , I think it' time to star talking to @BrunoCoimbra to check what xrootd client we build in COMP and to include that in the COMP code on cvmfs.

**Comment by amaltaro** [comment]
email sent to comp-ops.

**Comment by amaltaro** [comment]
@hufnagel xrootd is now deployed in cvmfs comp (see Bockjoo's reply).
Can you update this PR such that I can test it tomorrow and hopefully cut a new production tag before the weekend? Thanks

**Comment by hufnagel** [comment]
Updated but completely untested. Also a bigger update than expected maybe since not all sites might use cvmfs. We still support software deployment locally and using VO_SW_DIRK or OSG_APP environment to point at it. Supporting that too.

**Comment by hufnagel** [comment]
Fixed a bug discovered by jenkins...

File: src/python/WMCore/Storage/Backends/XRDCPImpl.py [hunk 1]
```diff
@@ -77,12 +79,39 @@ def createStageOutCommand(self, sourcePFN, targetPFN, options=None, checksums=No
...
+                    cmsSoftDir = os.path.join(cmsSoftDir, "cmssoft/cms")
+                else:
+                    cmsSoftDir = os.environ.get("CVMFS", None)
+
+            if cmsSoftDir:
+
+                initFiles = []
+                initFiles.append(os.path.join(cmsSoftDir, "COMP/slc6_amd64_gcc493/external/xrootd/4.0.4-comp/etc/profile.d/init.sh"))
+                initFiles.append(os.path.join(cmsSoftDir, "COMP/slc6_amd64_gcc493/external/libevent/2.0.22/etc/profile.d/init.sh"))
+                initFiles.append(os.path.join(cmsSoftDir, "COMP/slc6_amd64_gcc493/external/gcc/4.9.3/etc/profile.d/init.sh"))
```
Comment (amaltaro) [inline src/python/WMCore/Storage/Backends/XRDCPImpl.py]: Do you need to source the xrootd env AND its dependencies? I think just the xrootd one should be enough.

Comment (hufnagel) [inline src/python/WMCore/Storage/Backends/XRDCPImpl.py]: It isn't, tested it.

**Comment by amaltaro** [comment]
Darn it. The first test fails at FNAL, though I didn't really get it since there is the mapping from the command to the WMCore implementation...
```
==>Working on file: /store/unmerged/logs/prod/2016/10/17/amaltaro_MonteCarlo_eff_Agent1021_Validation_161017_164352_2987/Production/ProductionMergeRAWSIMoutput/0000/0/081202da-9482-11e6-b15c-02163e00ed6f-1-0-logArchive.tar.gz
===> Attempting Local Stage In.
LFN to PFN match made:
LFN: /store/unmerged/logs/prod/2016/10/17/amaltaro_MonteCarlo_eff_Agent1021_Validation_161017_164352_2987/Production/ProductionMergeRAWSIMoutput/0000/0/081202da-9482-11e6-b15c-02163e00ed6f-1-0-logArchive.tar.gz
PFN: root://cmsdcadisk01.fnal.gov//dcache/uscmsdisk/store/unmerged/logs/prod/2016/10/17/amaltaro_MonteCarlo_eff_Agent1021_Validation_161017_164352_2987/Production/ProductionMergeRAWSIMoutput/0000/0/081202da-9482-11e6-b15c-02163e00ed6f-1-0-logArchive.tar.gz

===> Local Stage Out Failure for file:
======>  /store/unmerged/logs/prod/2016/10/17/amaltaro_MonteCarlo_eff_Agent1021_Validation_161017_164352_2987/Production/ProductionMergeRAWSIMoutput/0000/0/081202da-9482-11e6-b15c-02163e00ed6f-1-0-logArchive.tar.gz
StageOutFailure
Message: Unable to retrieve impl for local stage in:
Error retrieving StageOutImpl for command named: stageout-xrdcp-fnal

        ErrorCode : 60311
        ModuleName : WMCore.Storage.StageOutError
        MethodName : __init__
        LFN : /store/unmerged/logs/prod/2016/10/17/amaltaro_MonteCarlo_eff_Agent1021_Validation_161017_164352_2987/Production/ProductionMergeRAWSIMoutput/0000/0/081202da-9482-11e6-b15c-02163e00ed6f-1-0-logArchive.tar.gz
        ClassInstance : None
        FileName : /storage/local/data1/condor/execute/dir_2008353/glide_CapeR6/execute/dir_3474459/job/WMCore.zip/WMCore/Storage/StageOutError.py
        ClassName : None
        Command : stageout-xrdcp-fnal
        LineNumber : 32
        ErrorNr : 0
        ErrorType : GeneralStageOutFailure
        ExceptionDetail : 'NoneType' object is not callable

Traceback:
  File "/storage/local/data1/condor/execute/dir_2008353/glide_CapeR6/execute/dir_3474459/job/WMCore.zip/WMCore/Storage/StageInMgr.py", line 240, in localStageIn
    impl = retrieveStageOutImpl(command, stagein=True)

  File "/storage/local/data1/condor/execute/dir_2008353/glide_CapeR6/execute/dir_3474459/job/WMCore.zip/WMCore/Storage/Registry.py", line 80, in retrieveStageOutImpl
    return classRef(stagein)
```

**Comment by hufnagel** [comment]
How is that error relevant ? This is stageout failing at FNAL, which has nothing to do with this patch.

**Comment by amaltaro** [comment]
I always mix up stage in with stage out. I thought stage out was only used to write to the storage and stage in to read/copy from the storage to the local disk.

Anyways, the real error is fixed in #7302 . Running another battery of tests now.

**Comment by amaltaro** [comment]
Ok, patch seems to be working now. I ran jobs in a couple other T1s and everything looks Ok.
@ticoann please merge

**Pull request merged**